File: update_schema.py
import sys
import re

def main():
    file_path = 'prisma/schema.prisma'
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    # Insert into User
    if 'assignedOperations      ProductionOperation[]' not in content:
        content = re.sub(r'(model User \{.*?)(^\})', r'\1  assignedOperations      ProductionOperation[]  @relation("OperationAssignedUser")\n  operationLogs           ProductionOperationLog[] @relation("OperationLogActor")\n\2', content, flags=re.MULTILINE | re.DOTALL)

    # Insert into ProductionJob
    if 'operations              ProductionOperation[]' not in content and 'model ProductionJob {' in content:
        content = re.sub(r'(model ProductionJob \{.*?)(^\})', r'\1  operations              ProductionOperation[]\n\2', content, flags=re.MULTILINE | re.DOTALL)

    # Insert into PrintQueueItem
    if 'operations              ProductionOperation[]' not in content and 'model PrintQueueItem {' in content:
        content = re.sub(r'(model PrintQueueItem \{.*?)(^\})', r'\1  operations              ProductionOperation[]\n\2', content, flags=re.MULTILINE | re.DOTALL)

    # Insert into OrderItem
    if 'operations              ProductionOperation[]' not in content and 'model OrderItem {' in content:
        content = re.sub(r'(model OrderItem \{.*?)(^\})', r'\1  operations              ProductionOperation[]\n\2', content, flags=re.MULTILINE | re.DOTALL)

    # Insert into ProductionMachine
    if 'operations              ProductionOperation[]' not in content and 'model ProductionMachine {' in content:
        content = re.sub(r'(model ProductionMachine \{.*?)(^\})', r'\1  operations              ProductionOperation[]\n\2', content, flags=re.MULTILINE | re.DOTALL)

    # Append new models
    new_models = '''
model OperationDefinition {
  id              String   @id @default(uuid())
  code            String   @unique
  name            String
  defaultSequence Int
  isActive        Boolean  @default(true)
  requiresMachine Boolean  @default(false)
  allowOutsource  Boolean  @default(false)
  description     String?
  createdAt       DateTime @default(now())
  updatedAt       DateTime @updatedAt
}

model ProductionOperation {
  id                  String   @id @default(uuid())
  productionJobId     String
  printQueueItemId    String?
  orderItemId         String?
  operationCode       String
  operationName       String
  sequence            Int
  status              String   @default("WAITING_PREVIOUS")
  machineId           String?
  assignedToId        String?
  
  inputSheets         Int
  plannedSheets       Int
  completedSheets     Int      @default(0)
  goodSheets          Int      @default(0)
  wasteSheets         Int      @default(0)
  
  errorReason         String?
  pauseReason         String?
  outsourceVendorName String?
  outsourceSentAt     DateTime?
  outsourceExpectedReturnAt DateTime?
  outsourceReceivedAt DateTime?
  
  startedAt           DateTime?
  completedAt         DateTime?
  createdAt           DateTime @default(now())
  updatedAt           DateTime @updatedAt

  productionJob   ProductionJob     @relation(fields: [productionJobId], references: [id], onDelete: Cascade)
  printQueueItem  PrintQueueItem?   @relation(fields: [printQueueItemId], references: [id], onDelete: Cascade)
  orderItem       OrderItem?        @relation(fields: [orderItemId], references: [id], onDelete: Cascade)
  machine         ProductionMachine?@relation(fields: [machineId], references: [id])
  assignedTo      User?             @relation("OperationAssignedUser", fields: [assignedToId], references: [id])
  logs            ProductionOperationLog[]

  @@index([productionJobId])
  @@index([printQueueItemId])
  @@index([orderItemId])
  @@index([machineId])
  @@index([assignedToId])
  @@index([status])
}

model ProductionOperationLog {
  id                    String   @id @default(uuid())
  productionOperationId String
  actorId               String
  action                String
  fromStatus            String?
  toStatus              String?
  beforeData            String?
  afterData             String?
  note                  String?
  createdAt             DateTime @default(now())

  productionOperation ProductionOperation @relation(fields: [productionOperationId], references: [id], onDelete: Cascade)
  actor               User                @relation("OperationLogActor", fields: [actorId], references: [id])

  @@index([productionOperationId])
  @@index([actorId])
  @@index([createdAt])
}
'''
    if 'model OperationDefinition' not in content:
        content += new_models

    with open(file_path, 'w', encoding='utf-8') as f:
        f.write(content)
        
    print("Prisma schema updated successfully.")

if __name__ == '__main__':
    main()

File: update-schema.js
const fs = require('fs');
const path = require('path');

const schemaPath = path.join(__dirname, 'prisma', 'schema.prisma');
let schema = fs.readFileSync(schemaPath, 'utf8');

// Insert relations if not exists
if (!schema.includes('printQueueItems PrintQueueItem[]')) {
  schema = schema.replace(
    /model Order \{[\s\S]*?(?=\n\})/,
    (match) => match + '\n  printQueueItems        PrintQueueItem[]'
  );
  
  schema = schema.replace(
    /model ProductionJob \{[\s\S]*?(?=\n\})/,
    (match) => match + '\n  printQueueItems        PrintQueueItem[]'
  );

  schema = schema.replace(
    /model DesignFile \{[\s\S]*?(?=\n\})/,
    (match) => match + '\n  printQueueItems        PrintQueueItem[]'
  );

  schema = schema.replace(
    /model InventoryItem \{[\s\S]*?(?=\n\})/,
    (match) => match + '\n  printQueueItems        PrintQueueItem[]'
  );

  schema = schema.replace(
    /model User \{[\s\S]*?(?=\n\})/,
    (match) => match + '\n  assignedPrintJobs      PrintQueueItem[] @relation("PrintJobAssignedTo")\n  printQueueLogs         PrintQueueLog[]  @relation("PrintQueueLogActor")'
  );
  
  schema = schema.replace(
    /model InventoryReservation \{[\s\S]*?(?=\n\})/,
    (match) => match + '\n  printQueueItems        PrintQueueItem[]'
  );
}

const newModels = `
model ProductionMachine {
  id                        String   @id @default(uuid())
  machineCode               String   @unique
  machineName               String
  machineType               String   
  status                    String   @default("ACTIVE") 
  defaultSpeedSheetsPerHour Int?
  setupTimeMinutes          Int?
  maintenanceNote           String?
  createdAt                 DateTime @default(now())
  updatedAt                 DateTime @updatedAt
  
  printQueueItems           PrintQueueItem[]
}

model PrintQueueItem {
  id               String    @id @default(uuid())
  productionJobId  String
  orderId          String
  machineId        String?
  assignedToId     String?
  printFileId      String?
  materialId       String?
  inventoryReservationId String?
  
  sheetSize        String?
  labelsPerSheet   Int?
  finishedQuantity Int?
  estimatedSheets  Int?
  totalSheets      Int
  wasteSheets      Int?
  printedSheets    Int       @default(0)
  sheetCalculationNote String?

  numberOfSides    Int       @default(1)
  colorMode        String?
  printMode        String?
  
  status           String    @default("WAITING_ASSIGNMENT") 
  fileStatus       String    @default("NOT_CHECKED") 
  materialStatus   String    @default("NOT_CHECKED") 
  isMaterialReserved Boolean @default(false)
  reservedQuantity   Int     @default(0)
  
  priority         String    @default("NORMAL")
  queuePosition    Int       @default(0)
  scheduledDate    DateTime?
  deadline         DateTime?
  startAt          DateTime?
  endAt            DateTime?
  actualStartAt    DateTime?
  actualEndAt      DateTime?
  
  printNote        String?
  waitingReason    String?
  pauseReason      String?
  errorReason      String?
  blockReason      String?
  
  createdAt        DateTime  @default(now())
  updatedAt        DateTime  @updatedAt

  productionJob    ProductionJob     @relation(fields: [productionJobId], references: [id], onDelete: Cascade)
  order            Order             @relation(fields: [orderId], references: [id], onDelete: Cascade)
  machine          ProductionMachine? @relation(fields: [machineId], references: [id])
  assignedTo       User?             @relation("PrintJobAssignedTo", fields: [assignedToId], references: [id])
  printFile        DesignFile?       @relation(fields: [printFileId], references: [id])
  material         InventoryItem?    @relation(fields: [materialId], references: [id])
  inventoryReservation InventoryReservation? @relation(fields: [inventoryReservationId], references: [id])
  
  logs             PrintQueueLog[]

  @@index([machineId])
  @@index([status])
  @@index([productionJobId])
  @@index([orderId])
  @@index([scheduledDate])
}

model PrintQueueLog {
  id                  String   @id @default(uuid())
  printQueueItemId    String
  actorId             String
  action              String
  fromStatus          String?
  toStatus            String?
  printedSheetsBefore Int?
  printedSheetsAfter  Int?
  note                String?
  createdAt           DateTime @default(now())
  
  printQueueItem      PrintQueueItem @relation(fields: [printQueueItemId], references: [id], onDelete: Cascade)
  actor               User           @relation("PrintQueueLogActor", fields: [actorId], references: [id])
  
  @@index([printQueueItemId])
  @@index([actorId])
  @@index([createdAt])
}
`;

if (!schema.includes('model PrintQueueItem')) {
  schema += '\n' + newModels;
}

fs.writeFileSync(schemaPath, schema);
console.log('Schema updated successfully');
